File: android/reader/src/main/kotlin/io/github/a7asoft/nfcemv/reader/ReaderError.kt
package io.github.a7asoft.nfcemv.reader

import io.github.a7asoft.nfcemv.extract.EmvCardError
import io.github.a7asoft.nfcemv.extract.GpoError
import io.github.a7asoft.nfcemv.extract.PdolError
import io.github.a7asoft.nfcemv.extract.PpseError
import io.github.a7asoft.nfcemv.extract.SelectAidFciError

/**
 * Typed reasons a [ContactlessReader] read terminated in [ReaderState.Failed].
 *
 * Variants carry only structural metadata (status words, structural
 * sub-error references, [IoReason] enum). They never embed raw value
 * bytes from the card or arbitrary [Throwable] messages — the latter
 * would let the underlying exception's `message` string surface in
 * `toString()` and potentially leak PCI-sensitive context.
 */
public sealed interface ReaderError {

    /**
     * The IsoDep channel produced an `IOException`. The structural
     * [reason] enum identifies the subtype without embedding the raw
     * exception message.
     */
    public data class IoFailure(val reason: IoReason) : ReaderError

    /**
     * `SELECT 2PAY.SYS.DDF01` returned `6A 82` (file/application not
     * found). Card does not advertise a PPSE — fall back to PSE / direct
     * AID selection is out of scope for v0.2.0.
     */
    public data object PpseUnsupported : ReaderError

    /** PPSE returned no usable application templates. */
    public data object NoApplicationSelected : ReaderError

    /** A non-`90 00` status word was returned by an APDU we cannot recover from. */
    public data class ApduStatusError(val sw1: Byte, val sw2: Byte) : ReaderError

    /** PPSE FCI was malformed. */
    public data class PpseRejected(val cause: PpseError) : ReaderError

    /** GPO response was malformed. */
    public data class GpoRejected(val cause: GpoError) : ReaderError

    /** SELECT AID FCI body was malformed. */
    public data class SelectAidFciRejected(val cause: SelectAidFciError) : ReaderError

    /** PDOL bytes from SELECT AID FCI failed structural parse. */
    public data class PdolRejected(val cause: PdolError) : ReaderError

    /** `EmvParser.parse` rejected the assembled record stream. */
    public data class ParseFailed(val cause: EmvCardError) : ReaderError
}

/**
 * Structural classification of the IO failure surfaced by
 * [ReaderError.IoFailure]. Maps `android.nfc.tech.IsoDep`'s I/O
 * exception classes to a closed enum so [ReaderError.toString] cannot
 * leak whatever the underlying exception's message string carries.
 */
public enum class IoReason {
    /** `android.nfc.TagLostException` was thrown — tag moved out of field. */
    TagLost,

    /** RF transceive timed out without a response (e.g. `SocketTimeoutException`). */
    Timeout,

    /** Generic `java.io.IOException` without a more specific subtype. */
    Generic,
}

File: ios/Sources/EmvReader/ReaderError.swift
import Foundation
import Shared

/// Reasons ``EmvReader/read()`` can terminate with
/// ``ReaderState/failed(_:)``.
///
/// Mirrors the Kotlin `io.github.a7asoft.nfcemv.reader.ReaderError`
/// catalogue. Structural metadata only — never embeds raw card bytes
/// (CLAUDE.md §5.8). Kotlin sub-error protocols (`PpseError`,
/// `GpoError`, `EmvCardError`) cross the bridge as
/// `id<SharedXxxError>` and are passed through as `Any` here.
public enum ReaderError: @unchecked Sendable {
    /// CoreNFC's `NFCISO7816Tag` connection or transceive raised an
    /// `Error`. ``IoReason`` distinguishes tag-loss vs timeout vs
    /// generic failure without surfacing the underlying error's
    /// `localizedDescription` (which can carry implementation noise).
    case ioFailure(IoReason)

    /// PPSE returned `6A 82` (file not found) — the card has no
    /// contactless application directory. PSE (`1PAY.SYS.DDF01`)
    /// fallback is out of scope for v0.2.x.
    case ppseUnsupported

    /// A card APDU returned a non-`90 00` status word at a stage
    /// other than PPSE-not-found. Status-word bytes are non-sensitive
    /// (public protocol).
    case apduStatusError(sw1: UInt8, sw2: UInt8)

    /// PPSE response body had the `90 00` success status but the
    /// FCI structure failed `Ppse.parse`. The associated value is the
    /// Kotlin `PpseError` sealed interface boxed as `Any`.
    case ppseRejected(Any)

    /// GPO response body was `90 00` but failed `Gpo.parse`. The
    /// associated value is the Kotlin `GpoError` sealed interface
    /// boxed as `Any`.
    case gpoRejected(Any)

    /// READ RECORD records were collected but `EmvParser.parse`
    /// surfaced an error. The associated value is the Kotlin
    /// `EmvCardError` sealed interface boxed as `Any`.
    case parseFailed(Any)
}

/// Categorical I/O failure reason for ``ReaderError/ioFailure(_:)``.
///
/// Mirrors the Kotlin `IoReason` enum from the Android reader. Maps
/// CoreNFC's `NFCReaderError` codes to a small, stable set of
/// platform-neutral categories so consumers don't have to depend on
/// CoreNFC error code semantics.
public enum IoReason: Sendable {
    /// Maps to `NFCReaderError.readerTransceiveErrorTagConnectionLost`
    /// or any session invalidation caused by the tag leaving the
    /// reader's RF field.
    case tagLost

    /// Maps to session timeout invalidation
    /// (`NFCReaderError.readerSessionInvalidationErrorSessionTimeout`).
    case timeout

    /// Catch-all for any other CoreNFC error.
    case generic
}
